src/litserve/specs/openai.py

```diff
@@ -130,68 +126,80 @@ def encode_response(self, output: Union[Dict[str, str], List[Dict[str, str]]]) -
-                    asyncio.to_thread(self._server.get_from_pipe, read),
-                    self._server.timeout,
-                    uid,
-                    self._server.request_buffer,
-                )
+                data = self._server.win_data_streamer(read, write)
             else:
-                data = await wait_for_queue_timeout(
-                    self._server.data_reader(read), self._server.timeout, uid, self._server.request_buffer
-                )
-            responses.append(data)
-        return responses
-
-    async def chat_completion(
-        self, request: ChatCompletionRequest, background_tasks: BackgroundTasks
-    ) -> ChatCompletionResponse:
+                data = self._server.data_streamer(read, write)
+
+            choice_pipes.append(data)
+        return choice_pipes
+
+    async def chat_completion(self, request: ChatCompletionRequest, background_tasks: BackgroundTasks):
         logger.debug("Received chat completion request %s", request)
-        if request.stream:
-            raise HTTPException(400, "Stream is not supported")
 
         uids = [uuid.uuid4() for _ in range(request.n)]
         pipes = []
         for uid in uids:
             read, write = self._server.new_pipe()
-
             request_el = request.model_copy()
             request_el.n = 1
             self._server.request_buffer[uid] = (request_el, write)
             self._server.request_queue.put(uid)
-
             background_tasks.add_task(self._server.cleanup_request, self._server.request_buffer, uid)
             pipes.append((read, write))
 
-        responses = await self.get_from_pipe(uids, pipes)
-
+        responses = await self.get_from_pipes(uids, pipes)
         for read, write in pipes:
             self._server.dispose_pipe(read, write)
 
-        usage = UsageInfo()
+        if request.stream:
+            return StreamingResponse(self.streaming_completion(request, responses))
 
-        choices = []
-        for i, data in enumerate(responses):
-            response, status = data
-            logger.debug("Received chat completion response %s with status %s", response, status)
-            if status == LitAPIStatus.ERROR:
-                load_and_raise(response)
+        return await self.non_streaming_completion(request, responses)
 
-            if status != LitAPIStatus.OK:
+    async def streaming_completion(self, request: ChatCompletionRequest, pipe_responses: List):
+        model = request.model
+        usage = UsageInfo()
+        for i, response in enumerate(pipe_responses):
+            choices = []
+            async for choice in response:
+                choice = json.loads(choice)
+                logger.debug(choice)
+                choice = StreamingChoice(**choice)
+                choice.index = i
+                choices.append(choice)
+
+            yield ChatCompletionChunk(model=model, choices=choices, usage=usage, system_fingerprint="").json()
```

Comment:
Look at the spec: https://platform.openai.com/docs/api-reference/chat/streaming#chat/streaming-usage
By default usage is always `null` unless the message is the last message returned